### ISSUE TITLE
Multimaps.unmodifiableMultimap (and others) could be covariant

### ISSUE THREAD

**Issue by tavianator** [google/guava#2448]
Mutable Java collections can't be safely treated covariantly.  But there's a nice trick when you don't need to modify them:

``` java
void foo(List<Number> numbers);

void bar(List<Integer> ints) {
    foo(Collections.unmodifiableList(ints));
}
```

This works because `Collections.unmodifiableList()` takes in a `List<? extends T>` but returns a `List<T>`.  (It also requires Java 8's type inference; write `Collections.<Number>unmodifiableList(ints)` on Java &le; 7.)  Similarly for sets, maps, etc.

It would be nice to be able to do the same thing with `Multimap`s, but the `Multimaps.unmodifiable*` methods are not covariant.

**Comment by lowasser** [comment]
It sounds like this would have the side effect of breaking source compatibility with Java 7 users of these methods?

**Comment by cpovirk** [comment]
`unmodifiableMultimap` has some questionable behavior that makes this unsafe, albeit in unlikely situations. The `Multimap` it returns has a "smart" `get` method that will return you `SortedSet` if the delegate collection is a `SortedSet`. This gives you access to `comparator()`. As a result, if we let you turn a `Multimap<K, Integer>` into a `Multimap<K, Number>`, we also let you turn a `Comparator<Integer>` into a `Comparator<Number>`, and it might not work on arbitrary `Number` types.

Again, the questionable behavior lies in `unmodifiableMultimap`. If not for that, I *think* that this proposal would be safe. But as things are, it's not safe -- again, albeit only in unlikely situations. But given that the workaround, if you're willing to do unchecked conversions, is an unchecked cast, I don't think it's worth trying to change `unmodifiableMultimap` (which might break some users who depend on its "smart" `get` return types).

**Comment by tavianator** [comment]
@lowasser I don't think so, changing a method from `foo(A<B> a)` to `foo(A<? extends B> a)` shouldn't break any `foo()` calls.  Is there some corner case I'm missing?

@cpovirk Ah that's unfortunate.  At least `unmodifiableMultimap(Multimap<K, ? extends V>)` would be safe, but it's kind of weird to be covariant only in the values.  That's what `Collections.unmodifiableSortedMap()` does, presumably for similar reasons.

That whole feature is weird btw.  Even `unmodifiableSetMultimap()` doesn't dynamically return `SortedSet` proxies, only `unmodifiableMultimap()` does.  It looks like `unmodifiable{List,Set,SortedSet}Multimap()` could still be made fully covariant in keys and values.

**Comment by lowasser** [comment]
@tavianator I'm only referring to your post -- you said yourself "It also requires Java 8's type inference; write Collections.<Number>unmodifiableList(ints) on Java ≤ 7."  So wouldn't that imply that Java <= 7 users would be required to insert explicit type arguments?

**Comment by cpovirk** [comment]
Oddly enough, I think it's the *keys* that we can safely be covariant in: `keySet` and `keys` omit the problematic special casing for `Sorted*`.

<strike>Ha, actually, fun fact: The key methods *used* to have the special casing, too -- implicitly, because `transformValues` used to check `instanceof SortedMap`. But we removed that for unrelated reasons: 2d97fd8466f9bf00c7a90d1795b3ce0b891c030c That's arguably a bit of evidence that we could safely change the behavior of `unmodifiableMultimap(...).get(...)`.</strike> Wait, no: I'm not sure why I think `transformValues` is involved in `keys` or `keySet`. At most, it's involved in `asMap`, but there, we're wrapping the result of `transformValues` in an `unmodifiableMap`, so it ends up as a plain `Map` with no access to the key comparator.

Values is another story: That's where `get(...).comparator()` (and, I now see, `valueComparator()`) are available.

The reason that we're backward from `SortedMap` is that we offer `valueComparator()` but `SortedMap` offers [key]`Comparator`. It would make some sense for us to offer `keyComparator()`, too, but we were wary of the combinatorial explosion of `Multimap` types.

Good catch that `unmodifiableSortedMultimap` doesn't have this behavior. Very strange. I looked the the CL review, and it seems that the goal here might have been just to avoid casts, even at the loss of functionality (which the author acknowledged in a comment in the code).
